README.txt
getdirections module for Drupal 7.x

If you have any questions or suggestions please contact me at
http://drupal.org/user/52366 or use the Getdirections issue queue.

This module provides a form to make a Google directions map.

Installation
Upload the whole getdirections directory into sites/all/modules/ or
sites/<domain>/modules/ for multisite setups.

Login to Drupal site as a user with administrative rights and go to
Administer > Site building > Modules. Scroll down to Other modules section, you
should now see Getdirections module listed there. Tick it and save configuration.

After installation go to Administer > Site configuration > Get directions and
select your preferences and Save Configuration.

If you have the gmap module installed and configured, getdirections will
use its configuration as a starting point.

Also remember to go to Administer > User management > Permissions and
set up permissions according to your needs.

If you have the location or getlocations_fields module installed you can
make use of any nodes containing location information.

For instance, if you want to "preload" the getdirections form with information
about the destination use a URL in this format:

getdirections/location/to/99

Where '99' is the node id (nid) of the location.
The user will only have to fill in the starting point.

To do it the other way around use
getdirections/location/from/99

You can also get a map with waypoints
getdirections/locations_via/1,2,3,99

If you have both the starting point and destination node ids then you can use
getdirections/locations/1/99

Where '1' is the starting point node id and 99 is the destination node id
(note the 's' in locations)

If you are using the location_user or getlocations_fields module you can
make a link to a user's location with the user id:
getdirections/location_user/to/66

Same goes for 'from'

You can also get a map with waypoints of user locations with the user ids:
getdirections/locations_user_via/1,2,3,99

You can mix nodes and users with
getdirections/location_u2n/66/99
Where 66 is is user id, 99 is node id

and vice-versa
getdirections/location_n2u/99/66
Where 66 is is user id, 99 is node id

You can build a map by supplying latitude/longitude pairs delimited with a '|'
eg getdirections/latlons/nn,nn|nn,nn|nn,nn
where nn,nn is a latitude/longitude pair.
A maximum of 8 pairs is permitted by Google unless you have a commercial license.

If you have the Colorbox module installed and enabled in Get Directions
you can place any of the above paths in a colorbox iframe by replacing
'getdirections' with 'getdirections_box'.
To enable this for a link you can use the 'colorbox-load' method,
make sure that this feature has been enabled in colorbox
and use a url like this:
<a href="/getdirections_box?width=700&height=600&iframe=true" class="colorbox-load">See map</a>

or (advanced use) by adding rel="getdirectionsbox" to the url, eg
<a href="/getdirections_box" rel="getdirectionsbox">See map</a>

The last method uses the settings in admin/settings/getdirections for colorbox
and uses its own colorbox event handler, see getdirections.js. You can define your own
event handlers in your theme's javascript.
'getdirections_box' has it's own template, getdirections_box.tpl.php which can be
copied over to your theme's folder and tweaked there.

If you install
http://drupal.org/project/menu_attributes
you can get colorbox links into the menu system.
for instance, you can put
getdirections_box?width=700&height=600&iframe=true
in the path and
colorbox-load
in the class textfield
or if you have colorbox enabled in Get Directions:
getdirections_box
in the path and
getdirectionsbox
in the Relationship textfield.

------------------
Get Directions API
These functions are for use in other modules.
There are functions available to generate these paths:

getdirections_location_path($direction, $nid)
Where $direction is either 'to' or 'from' and $nid is the node id concerned.

getdirections_locations_path($fromnid, $tonid)
Where $fromnid is the starting point and $tonid is the destination

getdirections_locations_via_path($nids)
Where $nids is a comma delimited list, eg "1,2,3,99" of node ids
Google imposes a limit of 25 points

If you have the latitude and longitude of the start and end points you can use

getdirections_locations_bylatlon($fromlocs, $fromlatlon, $tolocs, $tolatlon)
Where $fromlocs is a description of the starting point,
$fromlatlon is the latitude,longitude of the starting point
and $tolocs and $tolatlon are the same thing for the endpoint

See getdirections.api.inc for more detail.

If you are using Views and Location it will provide a 'Get driving directions' block
when you are viewing a location node or a user with a location.

Getdirections now supports the Googlemaps API version 3, this has many new features.


The getdirections_latlon() function can be used in code or called by URL.
getdirections_latlon($direction, $latlon, $locs)
Where $direction is either 'to' or 'from', $latlon is the latitude,longitude to be used
and $locs is a string to describe the point being used.
The URL method would be in the form of
<a href="/getdirections/latlon/(from or to)/(lat,lon)/(description)">My Link</a>
You can also use the function getdirections_location_latlon_path() in getdirections.api.inc to
generate the string for you.

Themeing
To change the way the map is displayed you should copy the theme function you want to change
to your theme's template.php, renaming appropriately and editing it there.

There is plenty of help on themeing on drupal.org

An example of a theming change, from http://drupal.org/node/1113042

Your theme should have a template.php,
copy the function theme_getdirections_show() from getdirections.module to your template.php,
renaming it appropriately, eg from
theme_getdirections_show
to
mythemename_getdirections_show
where 'mythemename' is the name of your theme. Then change the following:

  $rows[] = array(
    array(
      'data' => '<div id="getdirections_map_canvas_' . $mapid . '" style="width: ' . $width . '; height: ' . $height . '" ></div>',
      'valign' => 'top',
      'align' => 'center',
      'class' => array('getdirections-map'),
    ),
    array(
      'data' => ($getdirections_defaults['use_advanced'] && $getdirections_defaults['advanced_alternate'] ? '<button id="getdirections-undo-' . $mapid . '">' . t('Undo') . '</button>' : '') . '<div id="getdirections_directions_' . $mapid . '"></div>',
      'valign' => 'top' ,
      'align' => 'left',
      'class' => array('getdirections-list'),
    ),
  );

to
  $rows[] = array(
    array(
      'data' => '<div id="getdirections_map_canvas_' . $mapid . '" style="width: ' . $width . '; height: ' . $height . '" ></div>',
      'valign' => 'top',
      'align' => 'center',
      'class' => array('getdirections-map'),
    ),
  );
  $rows[] = array(
    array(
      'data' => ($getdirections_defaults['use_advanced'] && $getdirections_defaults['advanced_alternate'] ? '<button id="getdirections-undo-' . $mapid . '">' . t('Undo') . '</button>' : '') . '<div id="getdirections_directions_' . $mapid . '"></div>',
      'valign' => 'top' ,
      'align' => 'left',
      'class' => array('getdirections-list'),
    ),
  );

What this does is rearrange the table so that it produces two rows with one datacell each instead of one row with two datacells.

Make sure you flush the theme registry when you have made the changes, the devel and admin_menu modules are helpful for this.

Putting Getdirections map and list into divs. This example should work with responsive themes.
replace everything from
  $header = array();
to
  $output .= '<div class="getdirections">' . theme('table', array('header' => $header, 'rows' => $rows)) . '</div>';

with
  $output .= '<div class="getdirections">';
  $output .= '<div class="getdirections-map"  style="width: ' . $width . '; height: ' . $height . '" >';
  $output .= '<div id="getdirections_map_canvas_' . $mapid . '" style="width: 100%; height: 100%" ></div>';
  $output .= '</div>';
  $output .= '<div class="getdirections-list">';
  if ($getdirections_defaults['use_advanced'] && $getdirections_defaults['advanced_alternate']) {
    $output .= '<button id="getdirections-undo-' . $mapid . '">' . t('Undo') . '</button>';
  }
  $output .= '<div id="getdirections_directions_' . $mapid . '"></div>';
  $output .= '</div>';
  $output .= '</div>';

Blocks.
Here is a simple way to crete a block with a form and map in it. Ensure you have php filter enabled and add
<?php
if (arg(0) == 'node') {
  $nid = arg(1);
  if (is_numeric($nid) && $nid > 0) {
    echo getdirections_setlocation('to', $nid);
  }
}
?>
This will only show on node view.

see http://drupal.org/node/880332

Adding a map directly into a block:
See http://drupal.org/node/1376392#comment-7135346

Create a custom block. Use the php filter for the content. Then paste this into your block content:

<?php
$n = FALSE;
$nid = FALSE;
if (arg(0)) {
  $n = arg(0);
}
if (arg(1)) {
$nid = arg(1);
}
if ($n && $nid && is_numeric($nid) && $nid > 0) {
  $l = getdirections_load_locations($nid, $n);
  if ($l) {
    echo getdirections_entity_setlocation($n ,'to', $nid);
  }
}
?>



CHANGELOG.txt
commit 193d3a3
Author: hutch <hutch@arwystli.net>
Date:   Sat May 7 18:01:31 2016 +0100

    applied define 'LANGUAGE_NONE'

getdirections.module

commit fd15fc4
Author: hutch <hutch@arwystli.net>
Date:   Tue Mar 1 16:34:44 2016 +0000

    added LICENSE.txt

CHANGELOG.txt
LICENSE.txt

commit d5ac95b
Author: hutch <hutch@arwystli.net>
Date:   Tue Mar 1 16:34:22 2016 +0000

    better api3, removed weather and panoramio

getdirections.module

commit 8a8255f
Author: hutch <hutch@arwystli.net>
Date:   Tue Feb 9 15:30:37 2016 +0000

    added signature and signed_in, removed weather and panoramio

CHANGELOG.txt

commit 1dc50d5
Author: hutch <hutch@arwystli.net>
Date:   Tue Feb 9 15:28:27 2016 +0000

    added signature and signed_in, removed weather and panoramio

getdirections.admin.inc
getdirections.install
getdirections.module
js/getdirections_admin.js
js/getdirections_preview.js
js/getdirections_v3.js
js/getdirections_v3a.js
js/getdirections_v3b.js
modules/getdirections_fields/js/getdirections_fields_formatter.js

commit c951225
Author: hutch <hutch@midwales.net>
Date:   Sun Oct 12 21:36:42 2014 +0100

    edited README

README.txt

commit bc8a6e6
Author: hutch <hutch@midwales.net>
Date:   Thu Sep 25 18:33:22 2014 +0100

    fixed unwanted check_plain, #2345015

CHANGELOG.txt
modules/getdirections_fields/getdirections_fields.module

commit e6c25ca
Author: hutch <hutch@midwales.net>
Date:   Thu Sep 25 18:32:25 2014 +0100

    added javascript load check

getdirections.module

commit 3a21570
Author: hutch <hutch@midwales.net>
Date:   Sat Jul 12 20:23:22 2014 +0100

    added uselatlons option

CHANGELOG.txt
getdirections.module
js/getdirections_v3.js
js/getdirections_v3a.js
js/getdirections_v3b.js

commit f4f4b4b
Author: hutch <hutch@midwales.net>
Date:   Sat Jul 12 19:56:34 2014 +0100

    only show installed message when using ui

getdirections.install

commit ef563d8
Author: hutch <hutch@midwales.net>
Date:   Sat Jul 12 19:41:09 2014 +0100

    added avoidFerries

getdirections.module
js/getdirections_v3.js
js/getdirections_v3a.js
js/getdirections_v3b.js

commit ea9395a
Author: hutch <hutch@midwales.net>
Date:   Thu May 22 13:06:39 2014 +0100

    added region_bias

CHANGELOG.txt
getdirections.module

commit 672ef40
Author: hutch <hutch@midwales.net>
Date:   Tue May 20 18:03:51 2014 +0100

    removed visual_refresh, improved https detection

CHANGELOG.txt
getdirections.module
js/getdirections_preview.js
js/getdirections_v3.js
js/getdirections_v3a.js
js/getdirections_v3b.js

commit c894a4f
Author: hutch <hutch@midwales.net>
Date:   Fri May 9 15:43:51 2014 +0100

    more via fix

CHANGELOG.txt
getdirections.module

commit d77603a
Author: hutch <hutch@midwales.net>
Date:   Fri May 9 10:13:10 2014 +0100

    bug in getdirections_entity_setlocations_via

CHANGELOG.txt
getdirections.module

commit 06b281d
Author: hutch <hutch@midwales.net>
Date:   Mon Apr 28 22:22:15 2014 +0100

    added vbo support

CHANGELOG.txt
modules/getdirections_fields/getdirections_fields.module

commit 7219401
Author: hutch <hutch@midwales.net>
Date:   Sun Mar 23 20:31:31 2014 +0000

    added direction option to map display

CHANGELOG.txt
modules/getdirections_fields/getdirections_fields.module

commit 17ce008
Author: hutch <hutch@midwales.net>
Date:   Mon Jan 27 18:18:34 2014 +0000

    fixed jquery bug, see #2182815

CHANGELOG.txt
js/getdirections_v3.js

commit dfc9af0
Author: hutch <hutch@midwales.net>
Date:   Sun Jan 12 23:14:43 2014 +0000

    fixed bug in class array, see #2172041

CHANGELOG.txt
README.txt
getdirections.module
modules/getdirections_fields/getdirections_fields.module

commit 946d020
Author: hutch <hutch@midwales.net>
Date:   Thu Jan 2 09:49:06 2014 +0000

    added function getdirections_setlocation for backward compat.

CHANGELOG.txt
getdirections.module

commit 6086e0f
Author: hutch <hutch@midwales.net>
Date:   Mon Nov 4 08:33:47 2013 +0000

    updated README.txt as in #1113042

README.txt

commit f43604c
Author: hutch <hutch@midwales.net>
Date:   Sat Oct 26 15:26:46 2013 +0100

    better block example in README

CHANGELOG.txt
README.txt

commit 22c9279
Author: hutch <hutch@midwales.net>
Date:   Sat Oct 12 07:42:27 2013 +0100

    cleaned up

modules/getdirections_fields/js/getdirections_fields_formatter.js

commit 44562af
Author: hutch <hutch@midwales.net>
Date:   Fri Oct 11 17:08:33 2013 +0100

    fixed jquery in formatter form

CHANGELOG.txt
js/getdirections_admin.js
modules/getdirections_fields/getdirections_fields.module
modules/getdirections_fields/js/getdirections_fields_formatter.js

commit 2e05a7a
Author: hutch <hutch@midwales.net>
Date:   Sat Oct 5 09:56:01 2013 +0100

    updated css

CHANGELOG.txt
getdirections.css

commit b53432e
Author: hutch <hutch@midwales.net>
Date:   Sat Oct 5 09:55:40 2013 +0100

    added descpription to menu item

getdirections.module

commit 2584861
Author: hutch <hutch@midwales.net>
Date:   Sat Oct 5 09:54:24 2013 +0100

    removed some tokens

getdirections.views_default.inc

commit ba5c951
Author: hutch <hutch@midwales.net>
Date:   Sat Oct 5 09:53:03 2013 +0100

    added theming example

README.txt

commit a3b5456
Author: hutch <hutch@midwales.net>
Date:   Mon Aug 26 11:14:06 2013 +0100

    added better google places check

CHANGELOG.txt
getdirections.module

commit 9c778e7
Author: hutch <hutch@midwales.net>
Date:   Mon Aug 19 12:56:58 2013 +0100

    separating translation strings

CHANGELOG.txt
getdirections.admin.inc
getdirections.module

commit b12bab8
Author: hutch <hutch@midwales.net>
Date:   Sun Aug 18 17:15:01 2013 +0100

    added nokeyboard and nodoubleclickzoom features

getdirections.module
js/getdirections_preview.js
js/getdirections_v3.js
js/getdirections_v3a.js
js/getdirections_v3b.js

commit 68aca4a
Author: hutch <hutch@midwales.net>
Date:   Sat Aug 17 15:19:16 2013 +0100

    google libraries compat with getlocations

getdirections.module

commit 3922890
Author: hutch <hutch@midwales.net>

getdirections_v2a.js

/**
 * @file
 * Javascript functions for getdirections module
 *
 * @author Bob Hutchinson http://drupal.org/user/52366
 * code derived from gmap_direx module
 * this is for googlemaps API version 2
 * with adaptations from econym.org.uk
*/
(function ($) {
  Drupal.getdirections = {};

  // errorcodes
  var reasons = [];
  reasons[G_GEO_SUCCESS]             = "Success";
  reasons[G_GEO_MISSING_ADDRESS]     = "Missing Address: The address was either missing or had no value.";
  reasons[G_GEO_UNKNOWN_ADDRESS]     = "Unknown Address:  No corresponding geographic location could be found for the specified address.";
  reasons[G_GEO_UNAVAILABLE_ADDRESS] = "Unavailable Address:  The geocode for the given address cannot be returned due to legal or contractual reasons.";
  reasons[G_GEO_BAD_KEY]             = "Bad Key: The API key is either invalid or does not match the domain for which it was given";
  reasons[G_GEO_TOO_MANY_QUERIES]    = "Too Many Queries: The daily geocoding quota for this site has been exceeded.";
  reasons[G_GEO_SERVER_ERROR]        = "Server error: The geocoding request could not be successfully processed.";
  reasons[G_GEO_BAD_REQUEST]         = "A directions request could not be successfully parsed.";
  reasons[G_GEO_MISSING_QUERY]       = "No query was specified in the input.";
  reasons[G_GEO_UNKNOWN_DIRECTIONS]  = "The GDirections object could not compute directions between the points.";

  //var map;
  var gdir;
  var geo;
  var bounds;
  var map;
  var trafficInfo;
  var traffictoggleState = 1;

  var path = [];
  var active = [];
  var gmarkers = [];
  var addresses = [];

  var state = 0;
  var llpatt = /[0-9.\-],[0-9.\-]/;
  var fromdone = false;
  var todone = false;
  var mylocale = 'en';
  var waypoints = 0;
  var startpoint = 0;
  var endpoint = 1;
  var viaIconColor = '';
  var viaIconUrl = "http://labs.google.com/ridefinder/images/mm_20_";

  // functions
  Drupal.getdirections.mygetDirections = function() {
    var from;
    var to;
    var s;
    var i;
    if (addresses[startpoint]) {
      from = addresses[startpoint] + "@" + path[startpoint].toUrlValue(6);
    }
    else {
      from = path[startpoint].toUrlValue(6);
    }
    if (addresses[endpoint]) {
      to = addresses[endpoint] + "@" + path[endpoint].toUrlValue(6);
    }
    else {
      to = path[endpoint].toUrlValue(6);
    }
    for (i = waypoints; i > 0; i--) {
      if (active[i]) {
        to = path[i].toUrlValue(6) +" to: " + to;
        gmarkers[i].hide();
      }
    }
    s = "from: " + from + " to: " + to;
    gdir.load(s, {getPolyline:true, locale:mylocale });

    gmarkers[startpoint].hide();
    gmarkers[endpoint].hide();

  }

  // without location module
  function setDirectionsfromto(fromAddress, toAddress) {
    Drupal.getdirections.mygetDirections();
  }

  function setDirectionsvia(lls, locale) {
    var arr = lls.split('|');
    gdir.loadFromWaypoints(arr, { "locale": locale });
  }

  function onGDirectionsLoad(){
    // Use this function to access information about the latest load() results.
    // e.g.
    //document.getElementById("getdirections_info").innerHTML = gdir.getStatus().code;
    // and yada yada yada...
    if (Drupal.settings.getdirections.show_distance) {
      $("#getdirections_show_distance").html(Drupal.settings.getdirections.show_distance + ': ' + gdir.getDistance().html);
    }
    if (Drupal.settings.getdirections.show_duration) {
      $("#getdirections_show_duration").html(Drupal.settings.getdirections.show_duration + ': ' + gdir.getDuration().html);
    }
  }

  function initialize() {
    if (! GBrowserIsCompatible()) {
      return;
    }

    function handleState() {
      var e;
      var point;
      if (! todone) {
        e = $("input[name=to]").val();
        if (e && e.match(llpatt)) {
          arr = e.split(",");
          point = new GLatLng(arr[0], arr[1]);
          createMarker(point, endpoint, icon3);
          path[endpoint] = point;
          map.panTo(path[endpoint]);
          todone = true;
        }
      }

      if (! fromdone) {
        e = $("input[name=from]").val();
        if (e && e.match(llpatt)) {
          arr = e.split(",");
          point = new GLatLng(arr[0], arr[1]);
          createMarker(point, startpoint, icon1);
          path[startpoint] = point;
          map.panTo(path[startpoint]);
          fromdone = true;
        }
      }

      if (state == 0) {
        if (fromdone) {
          state = 1;
        }
        else {
          $("#getdirections_start").show();
          $("#getdirections_end").hide();
          $("#getdirections_btn").hide();
          $("#getdirections_help").hide();
        }
      }
      if (state == 1) {
        if (todone) {
          state = 2;
        }
        else {
          $("#getdirections_start").hide();
          $("#getdirections_end").show();
          $("#getdirections_btn").hide();
          $("#getdirections_help").hide();
        }
      }
      if (state == 2) {
        if (todone) {
          dovias();
          setendbounds();
        }
        $("#getdirections_start").hide();
        $("#getdirections_end").hide();
        $("#getdirections_btn").show();
        $("#getdirections_nextbtn").hide();
        if (waypoints) {
          $("#getdirections_help").show();
        }
      }
    } // end handleState

    function createMarker(point, i, ic) {
      var t;
      if (i == startpoint) { t = 'From'; }
      else if (i == endpoint) { t = 'To'; }
      else { t = 'via ' + i; }
      var marker = new GMarker(point, {draggable:true, icon:ic, title:t});
      gmarkers[i] = marker;
      GEvent.addListener(marker, "dragend", function() {
        path[i] = marker.getPoint();
        map.panTo(path[i]);
        if (! active[i]) {
          swapMarkers(i);
          active[i] = true;
        }
        addresses[i] = "";
      });
      map.addOverlay(marker);
    }

    function swapMarkers(i) {
      map.removeOverlay(gmarkers[i]);
      createMarker(path[i], i, icon2);
    }

    function doStart(point) {
      createMarker(point, startpoint, icon1);
      path[startpoint] = point;
      state = 1;
      handleState();
    }

    function doEnd(point) {
      createMarker(point, endpoint, icon3);
      path[endpoint] = point;
      state = 2;
      handleState();
      dovias();
      setendbounds();
      if (waypoints) {
        $("#getdirections_help").show();
      }
    }

    function dovias() {
      for (var i = 1; i < endpoint; i++) {
        var lat = (path[startpoint].lat()*(endpoint-i) + path[endpoint].lat()*i)/endpoint;
        var lng = (path[startpoint].lng()*(endpoint-i) + path[endpoint].lng()*i)/endpoint;
        var p = new GLatLng(lat,lng);
        createMarker(p, i, icon4);
        path[i] = p;
      }
    }

    function setendbounds() {
      bounds.extend(path[startpoint]);
      bounds.extend(path[endpoint]);
      map.setZoom(map.getBoundsZoomLevel(bounds));
      map.setCenter(bounds.getCenter());
    }

    // ====== Geocoding ======
    function showAddress() {
      var s;
      if (state == 0) {
        s = $("input[name=from]").val();
        if ($("#edit-country-from").val()) {
          s += ', ' + $("#edit-country-from").val();
        }
        addresses[startpoint] = s;
      }
      if (state == 1) {
        s = $("input[name=to]").val();
        if ($("#edit-country-to").val()) {
          s += ', ' + $("#edit-country-to").val();
        }
        addresses[endpoint] = s;
      }
      geo.getLatLng(s, function (point) {
        if (point) {
          if (state == 1) { doEnd(point); }
          if (state == 0) { doStart(point); }
        }
        else {
          var result = geo.getCache().get(s);
          if (result) {
            reason = "Code "+ result.Status.code;
            if (reasons[result.Status.code]) { reason = reasons[result.Status.code]; }
          }
          else {
            reason = "";
          }
          alert(Drupal.t('Could not find') + ' "' + s + '" ' + reason);
        }
      });
    }

    // start
    var lat = Drupal.settings.getdirections.lat;
    var lng = Drupal.settings.getdirections.lng;
    var zoom = Drupal.settings.getdirections.zoom;
    var controltype = Drupal.settings.getdirections.controltype;
    var mtc = Drupal.settings.getdirections.mtc;
    var scale = Drupal.settings.getdirections.scale;
    var overview = Drupal.settings.getdirections.overview;
    var googlebar = Drupal.settings.getdirections.googlebar;
    var maptype = (Drupal.settings.getdirections.maptype ? Drupal.settings.getdirections.maptype : '');
    var baselayers = (Drupal.settings.getdirections.baselayers ? Drupal.settings.getdirections.baselayers : '');

    var fromlatlon = (Drupal.settings.getdirections.fromlatlon ? Drupal.settings.getdirections.fromlatlon : '');
    var tolatlon = (Drupal.settings.getdirections.tolatlon ? Drupal.settings.getdirections.tolatlon : '');
    mylocale = (Drupal.settings.getdirections.mylocale ? Drupal.settings.getdirections.mylocale : 'en');
    // pipe delim
    var latlons = (Drupal.settings.getdirections.latlons ? Drupal.settings.getdirections.latlons : '');
    // color of 'via' icon, can be black brown green purple yellow blue gray orange red white
    viaIconColor = (Drupal.settings.getdirections.waypoint_color ? Drupal.settings.getdirections.waypoint_color : 'white');

    waypoints = (Drupal.settings.getdirections.waypoints ? Drupal.settings.getdirections.waypoints : 0);
    if (waypoints) {
      endpoint = waypoints + 1;
      active[startpoint] = true;
      for (var i = 1; i <= waypoints; i++) { active[i] = false; }
      active[endpoint] = true;
    }
    else {
      active[startpoint] = true;
      active[endpoint] = true;
    }

    var mapOptions = {};
    if (googlebar) {
      mapOptions = {
        googleBarOptions : {
          style : "new"
        }
      }
    }

    // make map
    map = new GMap2(document.getElementById("getdirections_map_canvas"), mapOptions);

    // menu type
    if (mtc == 'standard') { map.addControl(new GMapTypeControl()); }
    else if (mtc == 'hier') { map.addControl(new GHierarchicalMapTypeControl()); }
    else if (mtc == 'menu') { map.addControl(new GMenuMapTypeControl()); }
    // nav control type
    if (controltype == 'micro') { map.addControl(new GSmallZoomControl()); }
    else if (controltype == 'micro3D') { map.addControl(new GSmallZoomControl3D()); }
    else if (controltype == 'small') { map.addControl(new GSmallMapControl()); }
    else if (controltype == 'large') { map.addControl(new GLargeMapControl()); }
    else if (controltype == 'large3D') { map.addControl(new GLargeMapControl3D()); }
    if (baselayers.Physical) { map.addMapType(G_PHYSICAL_MAP); }
    // map type
    if (maptype) {
      if (maptype == 'Map' && baselayers.Map) { map.setMapType(G_NORMAL_MAP); }
      if (maptype == 'Satellite' && baselayers.Satellite) { map.setMapType(G_SATELLITE_MAP); }
      if (maptype == 'Hybrid' && baselayers.Hybrid) { map.setMapType(G_HYBRID_MAP); }
      if (maptype == 'Physical' && baselayers.Physical) { map.setMapType(G_PHYSICAL_MAP); }
    }
    //GScaleControl()
    if (scale) { map.addControl(new GScaleControl()); }
    //GOverviewMapControl()
    if (overview) { map.addControl(new GOverviewMapControl()); }

    latlng = new GLatLng(lat, lng);
    map.setCenter(latlng, parseInt(zoom));

    if (googlebar) {
      map.enableGoogleBar();
    }

    if (Drupal.settings.getdirections.trafficinfo) {
      var trafficOptions = {incidents:true};
      trafficInfo = new GTrafficOverlay(trafficOptions);
      map.addOverlay(trafficInfo);
    }

    // some icons
    var baseIcon = new GIcon(G_DEFAULT_ICON);
    baseIcon.iconSize=new GSize(24,38);
    var icon1 = G_START_ICON;
    var icon2 = G_PAUSE_ICON;
    var icon3 = G_END_ICON;
    var icon4 = new GIcon(baseIcon, viaIconUrl + viaIconColor + ".png");
        icon4.shadow = viaIconUrl + "shadow.png";
        icon4.iconSize = new GSize(12, 20);
        icon4.shadowSize = new GSize(22, 20);
        icon4.iconAnchor = new GPoint(6, 20);
        icon4.infoWindowAnchor = new GPoint(5, 1);

    // Bounding
    bounds = new GLatLngBounds();

    // Create a Client Geocoder
    geo = new GClientGeocoder(new GGeocodeCache());

    handleState();

    GEvent.addListener(map, "click", function(overlay,point) {
      if (point) {
        if (state == 1) {
          doEnd(point);
        }
        if (state == 0) {
          doStart(point);
          if (! todone) {
            map.panTo(point);
          }
        }
      }
    });

    gdir = new GDirections(map, document.getElementById("getdirections_directions"));

    GEvent.addListener(gdir, "error", function() {
      var code = gdir.getStatus().code;
      var reason="Code " + code;
      if (reasons[code]) {
        reason = "Code " + code + " : " + reasons[code];
      }
      alert(Drupal.t('Failed to obtain directions') + ", " + reason);
    });

    GEvent.addListener(gdir, "load", onGDirectionsLoad);

    if (fromlatlon && tolatlon) {
      if (fromlatlon.match(llpatt) && tolatlon.match(llpatt)) {
          // we have a latlon, so get the points and run through mygetDirections().
          arr = fromlatlon.split(",");
          point = new GLatLng(arr[0], arr[1]);
          path[startpoint] = point;
          fromdone = true;

          arr = tolatlon.split(",");
          point = new GLatLng(arr[0], arr[1]);
          path[endpoint] = point;
          todone = true;
          state = 2;
          // no vias in this instance, no button to click
          Drupal.mygetDirections();
        }
    }


    if (latlons) {
      setDirectionsvia(latlons, mylocale);
    }

    // minding textfields
    $("input[name=from]").change( function() {
      showAddress();
    });
    $("input[name=to]").change( function() {
      showAddress();
    });

  } // end initialise

  Drupal.getdirections.nextbtn = function() {
    return;
  }

  Drupal.getdirections.toggleTraffic = function() {
    if (traffictoggleState == 1) {
      map.removeOverlay(trafficInfo);
      traffictoggleState = 0;
    }
    else {
      map.addOverlay(trafficInfo);
      traffictoggleState = 1;
    }
  }

  Drupal.behaviors.getdirections = {
    attach: function() {
      initialize();

      $(window).unload(function(){
        GUnload();
      });
    }
  };

})(jQuery);
